include/boost/url/decoded_view.hpp

```diff
@@ -0,0 +1,648 @@
+    only contain the const-qualified functions
+    `begin()`/`end()`.
+
+    Any operation on the underlying percent-encoded
+    strings invalidates the view iterators. The view
+    should be recreated after any operation that
+    may mutate the underlying range.
+
+*/
+class decoded_view
+{
+    char const* p_{nullptr};
+    std::size_t n_{0};
+    std::size_t dn_{0};
+    bool plus_to_space_{true};
+
+    using traits_type = std::char_traits<char>;
+
+public:
+    /** Type of a decoded character
+    */
+    using value_type = char;
+
+    /// @copydoc value_type
+    using reference = char;
+
+    /// @copydoc value_type
+    using const_reference = char;
+
+    /** The unsigned integer type used to represent size.
+    */
+    using size_type = std::size_t;
+
+    /** The signed integer type used to represent differences.
+    */
+    using difference_type = std::ptrdiff_t;
+
+    /** A read-only bidirectional iterator to the decoded range.
+
+        This is a read-only bidirectional iterator to
+        the decoded characters.
+
+    */
+#ifdef BOOST_URL_DOCS
+    using iterator = __see_below__;
+#else
+    class iterator;
+#endif
+
+    /// @copydoc const_iterator
+    using const_iterator = iterator;
+
+    /** A reverse read-only bidirectional iterator to the decoded range.
+
+        This is a read-only bidirectional iterator to
+        the decoded characters.
+
+    */
+    using reverse_iterator = std::reverse_iterator<iterator>;
+
+    /// @copydoc reverse_iterator
+    using const_reverse_iterator = std::reverse_iterator<const_iterator>;
+
+    /** Constructor
+
+        Default-constructed objects represent
+        the empty string.
+
+    */
+    decoded_view(
+        pct_decode_opts const& opt = {}) noexcept
+        : p_(nullptr)
+        , n_(0)
+        , dn_(0)
+        , plus_to_space_(opt.plus_to_space)
+    {}
+
+    /** Constructor
+    */
+    BOOST_URL_DECL
+    decoded_view(
+        string_view str,
+        pct_decode_opts opt = {}) noexcept;
+
+    /** Returns an iterator to the beginning
+    */
+    const_iterator
+    begin() const noexcept;
+
+    /** Returns an iterator to the end
+    */
+    const_iterator
+    end() const noexcept;
+
+    /** Accesses the first decoded character
+
+        @note The behavior is undefined if `empty() == true`.
+
+        @return reference to the first character
+    */
+    const_reference
+    front() const;
+
+    /** Accesses the last decoded character
+
+        @note The behavior is undefined if `empty() == true`.
+
+        @return reference to the first character
+    */
+    const_reference
+    back() const;
+
+    /** Returns the encoded string view
+
+        This function returns a string_view of
+        the underlying encoded string.
+
+        @return The encoded string view.
+    */
+    string_view
+    encoded() const
+    {
+        return {p_, n_};
+    }
+
+    /** Returns the number of char elements in the string
+    */
+    size_type
```

Comment:
if you want to declare this and the others like it `constexpr`, you could, and it would cost us nothing (as long as we don't have to use those ugly CONSTEXPR_CXX macros)